docs/docs/03-language-reference.md

```diff
@@ -572,20 +572,27 @@ The following is a list of supported indexable types:
 
 [`▲ top`][top]
 
-### 1.2 Utility Functions
-
-| Name         | Extra information                                     |
-| ------------ | ----------------------------------------------------- |
-| `log`        | logs str                                              |
-| `assert`     | checks a condition and _throws_ if evaluated to false |
-| `unsafeCast` | cast a value into a different type                    |
-| `nodeof`     | obtain the [tree node](./02-concepts/02-application-tree.md) of a preflight object |
-| `lift`       | explicitly qualify a [lift](./02-concepts/01-preflight-and-inflight.md#explicit-lift-qualification) of a preflight object |
+### 1.2 Intrinsic Functions
+
+Intrinsic functions are a special call-like expressions built into the Wing compiler with
+the following properties (given an example intrinsic `@x()`):
+- It is not possble to write `let variable = @x;` or otherwise use `@x` as an expression on its own
+- `x` is not automatically a symbol that can be referenced
+- The arguments/return types must be representable Wing types, but can be more dynamic than user-defined functions
+  - For example, the return type may change between inflight and preflight
+
+| Name            | Extra information                                                                                                         |
+| --------------- | ------------------------------------------------------------------------------------------------------------------------- |
+| `@log()`        | logs str                                                                                                                  |
+| `@assert()`     | checks a condition and _throws_ if evaluated to false                                                                     |
+| `@unsafeCast()` | cast a value into a different type                                                                                        |
+| `@nodeof()`     | obtain the [tree node](./02-concepts/02-application-tree.md) of a preflight object                                        |
+| `@lift()`       | explicitly qualify a [lift](./02-concepts/01-preflight-and-inflight.md#explicit-lift-qualification) of a preflight object |
```

Comment:
`@dirname` is another candidate.

I don't see a reason to limit this only to function-like builtins, right?

Reply:
I wanted to avoid listing possible candidates here. Any additions can hopefully be added as they're finalized.

> I don't see a reason to limit this only to function-like builtins, right?

I don't think there is a technical feasibility reason, but my thought is that only allowing `@x()` helps avoid thinking that @x is a "value". I don't feel super strongly about this so open to loosening it.

Reply:
`@dirname` is a value

Reply:
It doesn't have to be (`@dirname()`)

Also, per some other async discussion it's not going to be the initial approach for the relative path use case (coming soon)

Reply:
I'm remembering now why I had a more technical reason for avoiding just `@x`.
Let's say we have an intrinsic `@x` that returns a function. That means `@x()` would call the returned function. Does this mean any values passed as arguments there are part of the intrinsic call (where the compiler has special behaviors) or not? The parser could not handle this ambiguity so we'd have to decide dynamically.

Similar thing applies if it return an object or interface of some kind, where does the "intrinsic" part of `@x.func()` end? You could say it ends at `@x`, but then why would something like `@inflight("file.ts")` behave differently (where the argument passed and the entire expression get treated differently)?

Reply:
I think supporting "value intrinsics" (intrinsics that don't need to be called) makes sense.

re: implementation -- let's give another try to get this working. Maybe we can try parsing intrinsics as a new expression variant in tree-sitter, with a grammar rule like

```
intrinsic := "@" identifier ( "(" arg_list ")" )?
```

Or, we could leave the grammar as is, and add some logic to `parser.rs` that turns the appropriate call expression or identifier expression into a dedicated IntrinsicCall or IntrinsicValue node.

Or, we could augment the tree-sitter grammar to allow identifiers to start with "@", and just add appropriate checks in the compiler so you can't define new symbols starting with "@", and associate the built-in intrinsic symbols in the type checker with a special "intrinsic" Type.

Reply:
I have an approach that seems to work so I'll update this RFC to remove the restriction. For now it is basically the first approach you mention, to avoid `@` and intrinsic getting possibly mixed up with other expressions/references.